main/lsp/ExtractVariable.cc

```diff
@@ -131,110 +230,31 @@ class ExtractVariableWalk {
-                "loc on node doesn't match the loc found in the tree walk");
-        whereToInsert = methodDef->rhs.loc();
-    } else if (auto if_ = ast::cast_tree<ast::If>(*enclosingScope)) {
-        if (if_->thenp.loc().exists() && if_->thenp.loc().contains(locOffsets)) {
-            ENFORCE(!ast::isa_tree<ast::InsSeq>(if_->thenp));
-            whereToInsert = if_->thenp.loc();
-            ENFORCE(if_->thenp.loc() == extractVariableWalk.enclosingScopeLoc,
-                    "loc on node doesn't match the loc found in the tree walk");
-        } else {
-            ENFORCE(!ast::isa_tree<ast::InsSeq>(if_->elsep));
-            whereToInsert = if_->elsep.loc();
-            ENFORCE(if_->elsep.loc() == extractVariableWalk.enclosingScopeLoc,
-                    "loc on node doesn't match the loc found in the tree walk");
-        }
-    } else if (auto rescue = ast::cast_tree<ast::Rescue>(*enclosingScope)) {
-        if (rescue->body.loc().exists() && rescue->body.loc().contains(locOffsets)) {
-            ENFORCE(!ast::isa_tree<ast::InsSeq>(rescue->body));
-            ENFORCE(rescue->body.loc() == extractVariableWalk.enclosingScopeLoc,
-                    "loc on node doesn't match the loc found in the tree walk");
-            whereToInsert = rescue->body.loc();
-        } else if (rescue->else_.loc().exists() && rescue->else_.loc().contains(locOffsets)) {
-            ENFORCE(!ast::isa_tree<ast::InsSeq>(rescue->else_));
-            ENFORCE(rescue->else_.loc() == extractVariableWalk.enclosingScopeLoc,
-                    "loc on node doesn't match the loc found in the tree walk");
-            whereToInsert = rescue->else_.loc();
-        } else if (rescue->ensure.loc().exists() && rescue->ensure.loc().contains(locOffsets)) {
-            ENFORCE(!ast::isa_tree<ast::InsSeq>(rescue->ensure));
-            ENFORCE(rescue->ensure.loc() == extractVariableWalk.enclosingScopeLoc,
-                    "loc on node doesn't match the loc found in the tree walk");
-            whereToInsert = rescue->ensure.loc();
-        } else {
-            ENFORCE(false)
-        }
-    } else if (auto rescueCase = ast::cast_tree<ast::RescueCase>(*enclosingScope)) {
-        ENFORCE(!ast::isa_tree<ast::InsSeq>(rescueCase->body));
-        ENFORCE(rescueCase->body.loc() == extractVariableWalk.enclosingScopeLoc,
-                "loc on node doesn't match the loc found in the tree walk");
-        whereToInsert = rescueCase->body.loc();
-    } else if (auto while_ = ast::cast_tree<ast::While>(*enclosingScope)) {
-        ENFORCE(!ast::isa_tree<ast::InsSeq>(while_->body));
-        ENFORCE(while_->body.loc() == extractVariableWalk.enclosingScopeLoc,
-                "loc on node doesn't match the loc found in the tree walk");
-        whereToInsert = while_->body.loc();
+    auto enclosingScope = walk.enclosingScope;
+    auto whereToInsert = findWhereToInsert(*enclosingScope, locOffsets);
+    matchingNode = walk.matchingNode->deepCopy();
```

Comment:
I remain convinced it should be possible to remove this copy, but I think the change is in a good enough state for the time being and we can circle back to deal with this.